File: kasa/lib/models/household_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class HouseholdModel {
  final String id;
  final String name;
  final String inviteCode;
  final String currency;
  final String secondaryCurrency;
  final double exchangeRate;
  final String ownerId;
  final List<String> memberIds;
  final DateTime createdAt;

  HouseholdModel({
    required this.id,
    required this.name,
    required this.inviteCode,
    this.currency = 'USD',
    this.secondaryCurrency = 'CDF',
    this.exchangeRate = 2250.0,
    required this.ownerId,
    this.memberIds = const [],
    required this.createdAt,
  });

  factory HouseholdModel.fromFirestore(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return HouseholdModel(
      id: doc.id,
      name: data['name'] ?? '',
      inviteCode: data['inviteCode'] ?? '',
      currency: data['currency'] ?? 'USD',
      secondaryCurrency: data['secondaryCurrency'] ?? 'CDF',
      exchangeRate: (data['exchangeRate'] ?? 2250).toDouble(),
      ownerId: data['ownerId'] ?? '',
      memberIds: List<String>.from(data['memberIds'] ?? []),
      createdAt: (data['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
    );
  }

  Map<String, dynamic> toFirestore() {
    return {
      'name': name,
      'inviteCode': inviteCode,
      'currency': currency,
      'secondaryCurrency': secondaryCurrency,
      'exchangeRate': exchangeRate,
      'ownerId': ownerId,
      'memberIds': memberIds,
      'createdAt': Timestamp.fromDate(createdAt),
    };
  }

  String get formattedInviteCode => 'KASA-$inviteCode';

  HouseholdModel copyWith({
    String? id,
    String? name,
    String? inviteCode,
    String? currency,
    String? secondaryCurrency,
    double? exchangeRate,
    String? ownerId,
    List<String>? memberIds,
    DateTime? createdAt,
  }) {
    return HouseholdModel(
      id: id ?? this.id,
      name: name ?? this.name,
      inviteCode: inviteCode ?? this.inviteCode,
      currency: currency ?? this.currency,
      secondaryCurrency: secondaryCurrency ?? this.secondaryCurrency,
      exchangeRate: exchangeRate ?? this.exchangeRate,
      ownerId: ownerId ?? this.ownerId,
      memberIds: memberIds ?? this.memberIds,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

File: kasa_android/app/src/main/java/com/example/kasa/data/model/HouseholdModel.kt
package com.example.kasa.data.model

import java.util.Date

data class HouseholdModel(
    val id: String = "",
    val name: String = "",
    val inviteCode: String = "",
    val currency: String = "USD",
    val secondaryCurrency: String = "CDF",
    val exchangeRate: Double = 2250.0,
    val ownerId: String = "",
    val homeLatitude: Double? = null,
    val homeLongitude: Double? = null,
    val homeAddress: String = "",
    val memberIds: List<String> = emptyList(),
    val monthlyBudget: Double = 0.0,
    val createdAt: Date = Date(),
    val appIcon: String = "default",
    val homeRadiusMeters: Double = 150.0
) {
    val formattedInviteCode: String get() = "UNK-$inviteCode"
    val hasHomeLocation: Boolean get() = homeLatitude != null && homeLongitude != null
    fun isOwner(userId: String): Boolean = ownerId == userId

    fun isWithinHome(lat: Double, lng: Double): Boolean {
        if (!hasHomeLocation || homeLatitude == null || homeLongitude == null) return false
        val results = FloatArray(1)
        android.location.Location.distanceBetween(lat, lng, homeLatitude, homeLongitude, results)
        return results[0] <= homeRadiusMeters
    }
}
